src/vm/crossloaderallocatorhash.h

```diff
@@ -0,0 +1,188 @@
+//    then the memory efficiency per key trends toward 1.3333 * sizeof(Value). Otherwise basic
+//    cost per key/value pair (assuming they are pointer sized has an overhead of about 4 
+//    pointers + key/value data size.)
+template <class TRAITS>
+class CrossLoaderAllocatorHash
+{
+private:
+    typedef typename TRAITS::TKey TKey;
+    typedef typename TRAITS::TValue TValue;
+    typedef GCHeapHash<KeyToValuesGCHeapHashTraits<TRAITS>> KeyToValuesGCHeapHash;
+
+public:
+
+#ifndef DACCESS_COMPILE
+    // Add an entry to the CrossLoaderAllocatorHash, the default implementation of does DefaultCrossLoaderAllocatorHashTraits will not check for duplicates.
+    void Add(TKey key, TValue value, LoaderAllocator *pLoaderAllocatorOfValue);
+
+    // Remove an entry to the CrossLoaderAllocatorHash, only removes one entry
+    void Remove(TKey key, TValue value, LoaderAllocator *pLoaderAllocatorOfValue);
+
+    // Remove all entries that can be looked up by key
+    void RemoveAll(TKey key);
+#endif
+
+    // Using visitor walk all values associated with a given key. The visitor
+    // is expected to implement bool operator ()(OBJECTREF keepAlive, TKey key, TValue value).
+    // Return false from that function to stop visitation.
+    // This can be done simply by utilizing a lambda, or if a lambda cannot be used, a functor will do.
+    // The value of "value" in this case must not escape from the visitor object
+    // unless the keepAlive OBJECTREF is also kept alive
+    template <class Visitor>
+    bool VisitValuesOfKey(TKey key, Visitor &visitor);
+
+    // Visit all key/value pairs
+    template <class Visitor>
+    bool VisitAllKeyValuePairs(Visitor &visitor);
+
+    // Initialize this CrossLoaderAllocatorHash to be associated with a specific LoaderAllocator
+    // Must be called before any use of Add
+    void Init(LoaderAllocator *pAssociatedLoaderAllocator);
+
+private:
+#ifndef DACCESS_COMPILE
+    void EnsureManagedObjectsInitted();
+    LAHASHDEPENDENTHASHTRACKERREF GetDependentTrackerForLoaderAllocator(LoaderAllocator* pLoaderAllocator);
+    GCHEAPHASHOBJECTREF GetKeyToValueCrossLAHashForHashkeyToTrackers(LAHASHKEYTOTRACKERSREF hashKeyToTrackersUnsafe, LoaderAllocator* pValueLoaderAllocator);
+#endif // !DACCESS_COMPILE
+    
+    template <class Visitor>
+    static bool VisitKeyValueStore(OBJECTREF *pLoaderAllocatorRef, OBJECTREF *pKeyValueStore, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitTrackerAllEntries(LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitKeyToTrackerAllEntries(OBJECTREF hashKeyEntryUnsafe, Visitor &visitor);
+    static void DeleteEntryTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe);
+
+private:
+    LoaderAllocator *_loaderAllocator = 0;
```

Comment:
A nit - _loaderAllocator -> _pLoaderAllocator to make it consistent with the naming of function arguments